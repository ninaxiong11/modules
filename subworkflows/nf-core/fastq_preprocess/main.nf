include { FASTQ_SANITISE_SEQKIT } from '../fastq_sanitise_seqkit/main'
include { SEQKIT_SEQ            } from '../../../modules/nf-core/seqkit/seq/main'
include { SEQKIT_REPLACE        } from '../../../modules/nf-core/seqkit/replace/main'
include { SEQKIT_RMDUP          } from '../../../modules/nf-core/seqkit/rmdup/main'

workflow FASTQ_PREPROCESS {

    take:
    ch_reads              // channel: [ val(meta), [ fastq ] ]
    skip_seqkit_sana_pair // boolean
    skip_seqkit_seq       // boolean
    skip_seqkit_replace   // boolean
    skip_seqkit_rmdup     // boolean

    main:
    ch_versions = Channel.empty()

    if (!skip_seqkit_sana_pair) {
        FASTQ_SANITISE_SEQKIT( ch_reads )
        ch_reads        = FASTQ_SANITISE_SEQKIT.out.reads
        ch_versions     = ch_versions.mix(FASTQ_SANITISE_SEQKIT.out.versions.first())
    }

    // Split paired-end reads and add strandedness to meta
    ch_reads_split = ch_reads
        .flatMap { meta, reads ->
            if (meta.single_end) {
                if (reads instanceof List && reads.size() != 1) {
                    error("Error: Check your meta.single_end value. Single-end reads should contain one file only.")
                }
                return [[ meta + [strandness: 'single'], reads ]]
            } else {
                if (!(reads instanceof List) || reads.size() != 2) {
                    error("Error: Check your meta.single_end value. Paired-end data should have exactly 2 files.")
                }
                return [
                    [ meta + [strandness: 'R1'], reads[0] ],
                    [ meta + [strandness: 'R2'], reads[1] ]
                ]
            }
        }

    if (!skip_seqkit_seq) {
        SEQKIT_SEQ( ch_reads_split )
        ch_reads_split = SEQKIT_SEQ.out.fastx
        ch_versions    = ch_versions.mix(SEQKIT_SEQ.out.versions.first())
    }

    if (!skip_seqkit_replace) {
        SEQKIT_REPLACE( ch_reads_split )
        ch_reads_split = SEQKIT_REPLACE.out.fastx
        ch_versions    = ch_versions.mix(SEQKIT_REPLACE.out.versions.first())
    }

    if (!skip_seqkit_rmdup) {
        SEQKIT_RMDUP( ch_reads_split )
        ch_reads_split = SEQKIT_RMDUP.out.fastx
        ch_versions    = ch_versions.mix(SEQKIT_RMDUP.out.versions.first())
    }

    ch_reads = ch_reads_split
        .map { meta, fastq ->
            // Remove strandness field from meta to merge back together
            def clean_meta = meta.findAll { key, value -> key != 'strandness' }
            return [ clean_meta, fastq ]
        }
        .groupTuple(by: 0)
        .map { meta, files ->
            if (meta.single_end) {
                return [ meta, files[0] ]
            } else {
                def sorted_files = files.flatten().sort { it.name }
                return [ meta, sorted_files ]
            }
        }

    emit:
    reads    = ch_reads    // channel: [ val(meta), [ fastq ] ]
    versions = ch_versions // channel: [ versions.yml ]

}
