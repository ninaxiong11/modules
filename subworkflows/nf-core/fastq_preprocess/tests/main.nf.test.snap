{
    "sarscov2 - fastq - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_stub",
                            "single_end": true
                        },
                        "test_stub_single.fasta:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,12eff6cdaef31b58e36df720f8d96331",
                    "versions.yml:md5,8352db48172cddc55e42f943cdee1b9a",
                    "versions.yml:md5,9afc673105901c4307e8bb20d367a3ce",
                    "versions.yml:md5,d4403f4032d4cbb377d61ed83712253e"
                ],
                "reads": [
                    [
                        {
                            "id": "test_stub",
                            "single_end": true
                        },
                        "test_stub_single.fasta:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,12eff6cdaef31b58e36df720f8d96331",
                    "versions.yml:md5,8352db48172cddc55e42f943cdee1b9a",
                    "versions.yml:md5,9afc673105901c4307e8bb20d367a3ce",
                    "versions.yml:md5,d4403f4032d4cbb377d61ed83712253e"
                ]
            },
            [
                {
                    "FASTQ_PREPROCESS:SEQKIT_REPLACE": {
                        "seqkit": "2.9.0"
                    }
                },
                {
                    "FASTQ_PREPROCESS:SEQKIT_SEQ": {
                        "seqkit": "v2.9.0"
                    }
                },
                {
                    "FASTQ_PREPROCESS:FASTQ_SANITISE_SEQKIT:SEQKIT_SANA": {
                        "seqkit": "2.10.1"
                    }
                },
                {
                    "FASTQ_PREPROCESS:SEQKIT_RMDUP": {
                        "seqkit": "v2.9.0"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.10.0"
        },
        "timestamp": "2025-10-29T13:17:10.915765"
    },
    "sarscov2 - fastq - single_end": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_single",
                            "single_end": true
                        },
                        "test_single_single.fasta:md5,c1dbd30da9ad368bb7bf5a117bb765b2"
                    ]
                ],
                "1": [
                    "versions.yml:md5,12eff6cdaef31b58e36df720f8d96331",
                    "versions.yml:md5,8352db48172cddc55e42f943cdee1b9a",
                    "versions.yml:md5,9afc673105901c4307e8bb20d367a3ce",
                    "versions.yml:md5,d4403f4032d4cbb377d61ed83712253e"
                ],
                "reads": [
                    [
                        {
                            "id": "test_single",
                            "single_end": true
                        },
                        "test_single_single.fasta:md5,c1dbd30da9ad368bb7bf5a117bb765b2"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,12eff6cdaef31b58e36df720f8d96331",
                    "versions.yml:md5,8352db48172cddc55e42f943cdee1b9a",
                    "versions.yml:md5,9afc673105901c4307e8bb20d367a3ce",
                    "versions.yml:md5,d4403f4032d4cbb377d61ed83712253e"
                ]
            },
            [
                {
                    "FASTQ_PREPROCESS:SEQKIT_REPLACE": {
                        "seqkit": "2.9.0"
                    }
                },
                {
                    "FASTQ_PREPROCESS:SEQKIT_SEQ": {
                        "seqkit": "v2.9.0"
                    }
                },
                {
                    "FASTQ_PREPROCESS:FASTQ_SANITISE_SEQKIT:SEQKIT_SANA": {
                        "seqkit": "2.10.1"
                    }
                },
                {
                    "FASTQ_PREPROCESS:SEQKIT_RMDUP": {
                        "seqkit": "v2.9.0"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.10.0"
        },
        "timestamp": "2025-10-29T13:16:51.34468"
    },
    "sarscov2 - fastq - paired_end": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_paired",
                            "single_end": false
                        },
                        [
                            "test_paired_R1.fasta:md5,c1dbd30da9ad368bb7bf5a117bb765b2",
                            "test_paired_R2.fasta:md5,e69b523dbbc543b86b078718b55db757"
                        ]
                    ]
                ],
                "1": [
                    "versions.yml:md5,12eff6cdaef31b58e36df720f8d96331",
                    "versions.yml:md5,8352db48172cddc55e42f943cdee1b9a",
                    "versions.yml:md5,9afc673105901c4307e8bb20d367a3ce",
                    "versions.yml:md5,d4403f4032d4cbb377d61ed83712253e"
                ],
                "reads": [
                    [
                        {
                            "id": "test_paired",
                            "single_end": false
                        },
                        [
                            "test_paired_R1.fasta:md5,c1dbd30da9ad368bb7bf5a117bb765b2",
                            "test_paired_R2.fasta:md5,e69b523dbbc543b86b078718b55db757"
                        ]
                    ]
                ],
                "versions": [
                    "versions.yml:md5,12eff6cdaef31b58e36df720f8d96331",
                    "versions.yml:md5,8352db48172cddc55e42f943cdee1b9a",
                    "versions.yml:md5,9afc673105901c4307e8bb20d367a3ce",
                    "versions.yml:md5,d4403f4032d4cbb377d61ed83712253e"
                ]
            },
            [
                {
                    "FASTQ_PREPROCESS:SEQKIT_REPLACE": {
                        "seqkit": "2.9.0"
                    }
                },
                {
                    "FASTQ_PREPROCESS:SEQKIT_SEQ": {
                        "seqkit": "v2.9.0"
                    }
                },
                {
                    "FASTQ_PREPROCESS:FASTQ_SANITISE_SEQKIT:SEQKIT_SANA": {
                        "seqkit": "2.10.1"
                    }
                },
                {
                    "FASTQ_PREPROCESS:SEQKIT_RMDUP": {
                        "seqkit": "v2.9.0"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.10.0"
        },
        "timestamp": "2025-10-29T13:16:58.073323"
    },
    "sarscov2 - fastq - both with single broken": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_both",
                            "single_end": false
                        },
                        [
                            "test_both_R1.fasta:md5,c1dbd30da9ad368bb7bf5a117bb765b2",
                            "test_both_R2.fasta:md5,e69b523dbbc543b86b078718b55db757"
                        ]
                    ],
                    [
                        {
                            "id": "test_both",
                            "single_end": true
                        },
                        "test_both_single.fasta:md5,8d33c2a78ed2a29919b44e64b1251d6d"
                    ]
                ],
                "1": [
                    "versions.yml:md5,12eff6cdaef31b58e36df720f8d96331",
                    "versions.yml:md5,8352db48172cddc55e42f943cdee1b9a",
                    "versions.yml:md5,9afc673105901c4307e8bb20d367a3ce",
                    "versions.yml:md5,d4403f4032d4cbb377d61ed83712253e"
                ],
                "reads": [
                    [
                        {
                            "id": "test_both",
                            "single_end": false
                        },
                        [
                            "test_both_R1.fasta:md5,c1dbd30da9ad368bb7bf5a117bb765b2",
                            "test_both_R2.fasta:md5,e69b523dbbc543b86b078718b55db757"
                        ]
                    ],
                    [
                        {
                            "id": "test_both",
                            "single_end": true
                        },
                        "test_both_single.fasta:md5,8d33c2a78ed2a29919b44e64b1251d6d"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,12eff6cdaef31b58e36df720f8d96331",
                    "versions.yml:md5,8352db48172cddc55e42f943cdee1b9a",
                    "versions.yml:md5,9afc673105901c4307e8bb20d367a3ce",
                    "versions.yml:md5,d4403f4032d4cbb377d61ed83712253e"
                ]
            },
            [
                {
                    "FASTQ_PREPROCESS:SEQKIT_REPLACE": {
                        "seqkit": "2.9.0"
                    }
                },
                {
                    "FASTQ_PREPROCESS:SEQKIT_SEQ": {
                        "seqkit": "v2.9.0"
                    }
                },
                {
                    "FASTQ_PREPROCESS:FASTQ_SANITISE_SEQKIT:SEQKIT_SANA": {
                        "seqkit": "2.10.1"
                    }
                },
                {
                    "FASTQ_PREPROCESS:SEQKIT_RMDUP": {
                        "seqkit": "v2.9.0"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.10.0"
        },
        "timestamp": "2025-10-29T13:17:05.485679"
    }
}